Mathlib/Data/Complex/Basic.lean

```diff
@@ -0,0 +1,1376 @@
+    mul_div_cancel_left _ (mul_ne_zero two_ne_zero i_ne_zero : 2 * i ≠ 0)]
+
+
+#align complex.im_eq_sub_conj Complex.im_eq_sub_conj
+
+/-! ### Absolute value -/
+
+
+namespace AbsTheory
+
+-- mathport name: abs
+-- We develop enough theory to bundle `abs` into an `absolute_value` before making things public;
+-- this is so there's not two versions of it hanging around.
+set_option quotPrecheck false
+local notation "abs" z => (normSq z).sqrt
+
+private theorem mul_self_abs (z : ℂ) : ((abs z) * abs z) = normSq z :=
+  Real.mul_self_sqrt (normSq_nonneg _)
+-- #align complex.abs_theory.mul_self_abs Complex.AbsTheory.mul_self_abs
+
+private theorem abs_nonneg' (z : ℂ) : 0 ≤ abs z :=
+  Real.sqrt_nonneg _
+-- #align complex.abs_theory.abs_nonneg' Complex.AbsTheory.abs_nonneg'
+
+theorem abs_conj (z : ℂ) : (abs conj z) = abs z := by simp
+#align complex.abs_theory.abs_conj Complex.AbsTheory.abs_conj
+
+private theorem abs_re_le_abs (z : ℂ) : |z.re| ≤ abs z := by
+  rw [mul_self_le_mul_self_iff (abs_nonneg z.re) (abs_nonneg' _), abs_mul_abs_self, mul_self_abs]
+  apply re_sq_le_normSq
+-- #align complex.abs_theory.abs_re_le_abs Complex.AbsTheory.abs_re_le_abs
+
+private theorem re_le_abs (z : ℂ) : z.re ≤ abs z :=
+  (abs_le.1 (abs_re_le_abs _)).2
+-- #align complex.abs_theory.re_le_abs complex.abs_theory.re_le_abs
+
+private theorem abs_mul (z w : ℂ) : (abs z * w) = (abs z) * abs w := by
+  rw [normSq_mul, Real.sqrt_mul (normSq_nonneg _)]
+-- #align complex.abs_theory.abs_mul complex.abs_theory.abs_mul
+
+private theorem abs_add (z w : ℂ) : (abs z + w) ≤ (abs z) + abs w :=
+  (mul_self_le_mul_self_iff (abs_nonneg' (z + w)) (add_nonneg (abs_nonneg' z) (abs_nonneg' w))).2 <|
+    by
+    rw [mul_self_abs, add_mul_self_eq, mul_self_abs, mul_self_abs, add_right_comm, normSq_add,
+      add_le_add_iff_left, mul_assoc, mul_le_mul_left (zero_lt_two' ℝ), ←
+      Real.sqrt_mul <| normSq_nonneg z, ← normSq_conj w, ← map_mul]
+    exact re_le_abs (z * conj w)
+-- #align complex.abs_theory.abs_add complex.abs_theory.abs_add
+
+/-- The complex absolute value function, defined as the square root of the norm squared. -/
+noncomputable def Complex.abs : AbsoluteValue ℂ ℝ
+    where
+  toFun x := abs x
+  map_mul' := abs_mul
+  nonneg' := abs_nonneg'
+  eq_zero' _ := (Real.sqrt_eq_zero <| normSq_nonneg _).trans normSq_eq_zero
+  add_le' := abs_add
+#align complex.abs Complex.AbsTheory.Complex.abs
+
+end AbsTheory
+
+-- Porting note: Added this to make the following work.
+-- Hope this is as intended. Also the name has been expanded.
+open Complex.AbsTheory
+
+theorem abs_def : (Complex.abs : ℂ → ℝ) = fun z => (normSq z).sqrt :=
+  rfl
+#align complex.abs_def Complex.abs_def
+
+theorem abs_apply {z : ℂ} : Complex.abs z = (normSq z).sqrt :=
+  rfl
+#align complex.abs_apply Complex.abs_apply
+
+-- Porting note: removed `norm_cast`
+@[simp]
+theorem abs_ofReal (r : ℝ) : Complex.abs r = |r| := by
```

Comment:
why? the linter doesn't complain

Reply:
May have complained before the fix and I did not notice it is fine.